### PR TITLE
Upgrade JS Libraries

### PR DESCRIPTION
**Pull request by biyeun**
https://dimagi-dev.atlassian.net/browse/SAAS-10882

##### SUMMARY
Fixes the following security vulnerabilities:
- Use of JavaScript Library with Known Vulnerability, Knockout 3.3.0
- Use of JavaScript Library with Known Vulnerability, Bootstrap 3.3.7
- Use of JavaScript Library with Known Vulnerability, jQuery 3.2.1
- Use of JavaScript Library with Known Vulnerability, DOMPurify 0.8.3

##### RISK ASSESSMENT / QA PLAN
This should be thoroughly QA'd before merged, but I'm worried about the time pressure associated with this!

Note that this PR has to be merged and submodule ref updated before merging this PR
https://github.com/dimagi/MediaUploader/pull/28

### CODE REVIEW

File: corehq/apps/hqwebapp/static/hqwebapp/js/main.js [hunk 1]
```diff
@@ -89,6 +89,18 @@ hqDefine('hqwebapp/js/main', [
     };
     ko.virtualElements.allowedBindings.runOnInit = true;
 
+    ko.bindingHandlers.allowDescendantBindings = {
+        // fixes an ssue where we try to apply bindings to a parent element
```
Comment (orangejenny) [inline corehq/apps/hqwebapp/static/hqwebapp/js/main.js:93]:
```suggestion
        // fixes an issue where we try to apply bindings to a parent element
```

File: corehq/apps/hqwebapp/static/hqwebapp/js/knockout_bindings.ko.js [hunk 2]
```diff
@@ -130,14 +130,14 @@ hqDefine("hqwebapp/js/knockout_bindings.ko", [
...
                     if (item !== undefined) {
-                        // this is voodoo to me, but I have to remove the ui item from its new position
-                        // and *not replace* it in its original position for all the foreach mechanisms to work correctly
-                        // I found this by trial and error
-                        ui.item.detach();
                         //remove the item and add it back in the right spot
                         if (newPosition >= 0) {
-                            list.remove(item);
-                            list.splice(newPosition, 0, item);
+                            var newList = _.filter(list(), function (i) {
```
Comment (orangejenny) [inline corehq/apps/hqwebapp/static/hqwebapp/js/knockout_bindings.ko.js:135]: Is this the same as `_.without(list(), item)`?

File: corehq/apps/app_manager/static/app_manager/js/forms/advanced/case_config_ui.js [hunk 3]
```diff
@@ -24,6 +24,7 @@ hqDefine('app_manager/js/forms/advanced/case_config_ui', function () {
                 $('.property-description').closest('.read-only').popover({
                     'trigger': 'hover',
                     'placement': 'bottom',
+                    'sanitize': false,
```
Comment (orangejenny) [inline corehq/apps/app_manager/static/app_manager/js/forms/advanced/case_config_ui.js:27]: Is this safe, given that the property description is user input?

**Review by orangejenny — COMMENTED** [review]
Looks great, I bet this was fun to do...The only possibly blocking question is whether or not it's okay to use `sanitize: false` on the case config ui popovers.

**Review by millerdev — COMMENTED** [review]
LGTM after Jenny's issues are fixed.

File: corehq/apps/app_manager/static/app_manager/js/forms/advanced/case_config_ui.js (hunk 3 above)
Comment (biyeun) [reply corehq/apps/app_manager/static/app_manager/js/forms/advanced/case_config_ui.js:27]: updated

**Comment by biyeun** [comment]
@orangejenny @millerdev added a few more changes

**Review by millerdev — APPROVED** [review]

**Review by orangejenny — APPROVED** [review]

**Comment by orangejenny** [comment]
JS test failure might be real
```
1) App Releases SavedApp should only make one request when downloading zip:
     TypeError: 'undefined' is not a function (evaluating 'e.getAttribute("type")')
```

**Comment by biyeun** [comment]
@orangejenny weird this test is passing locally when run in the browser

**Comment by biyeun** [comment]
Just a note that I discussed with Clayton and came to the conclusion to comment out `grunt mocha` from are tests for the time being so we can get these updates in. The tests are all running in the browser, but failing in the command line / travis due to changes in how `jQuery` interprets the `DOM` and `PhantomJS` (which I think is what `grunt` is using) being hugely out of sync with those changes and now unmaintained. Going to move to `chrome headless`/`karma` I think...

cc @millerdev

**Pull request merged**